### DISCUSSION TITLE
Generate Github self-hosted runner token

### DISCUSSION THREAD

**Discussion by AF250329** [community/community#35613] [Actions] [answered]
Hi

I'm attempting to create a Terraform-integrated script that will create and configure a Google Cloud VM that will install Github Runner as self-hosted. The repository is under my workplace's 'organization' and it is closed to the public. Everything goes smoothly until I need to configure the runner. In repository instructions for creating self-hosted runner written as this:

```powershell
# Create the runner and start the configuration experience
$ ./config.cmd --url https://github.com/my_work_place_organizaiton_name/repository_name --token ASZER2QS4UVEAL3YLMZ3DIMUIC
```

The issue is that, because it is an unattended script, it will run entirely on its own with no strings attached, and everything should be generated as automatically as possible. So I need a way to generate/retrieve this token `ASZER2QS4UVEAL3YLMZ3DIMUIC` automatically.

I think I found a way (correct me if I wrong) here: [Create a registration token for an organization](https://docs.github.com/en/rest/actions/self-hosted-runners#create-a-registration-token-for-an-organization). So far so good. I managed to create a powershell script to execute all steps in new Github self-hosted runner until the step where I need to generate token. Once I run the command (even in Github CLI) I get an error back like this:

```powershell
gh api --method POST -H "Accept: application/vnd.github+json" /orgs/my_work_place_organizaiton_name/actions/runners/registration-token
{
  "message": "Must have admin rights to Repository.",
  "documentation_url": "https://docs.github.com/rest/reference/actions#create-a-registration-token-for-an-organization"
}
gh: Must have admin rights to Repository. (HTTP 403)
gh: This API operation needs the "admin:org" scope. To request it, run:  gh auth refresh -h github.com -s admin:org
```

I am an admin in this repository but not in the organization, and I am afraid that no one will grant me admin access to the organization, and even more, I cannot simply put admin:org credentials in some script - this is a "no go."

So, my question is, how can I fully automate the generation of this Github token (which is generated for everyone in the instructions page without any admin privileges)?

**Comment by jsoref** [comment]
You can have runners for organizations or for repositories. 

https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners

I'll note that your API example has a typo – I'd suggest using 📋.

**Reply by AF250329** [reply]
Hi @jsoref ! 
I appreciate the answer but 1. Of course API url is redacted because it is private repo 2. The error is not in API but in user rights to generate the token. You see the error from Github Cli: `Must have admin rights to Repository. (HTTP 403)`. 

This is my question is about: How does everyone (or let say `admin`) in this repo could simple copy/paste the token from instructions page but to generate the token automatically I need `admin:org` access to the repo ? 
 
How to generate this token automatically without `admin:org` in Organization ?

**Reply by jsoref** [reply]
Oh, I was complaining about `organizaiton` instead of `organization`. -- Sorry, half the time these typos matter.

Anyway, you're using the `/orgs/` entrypoint, you need to use the `/repos/` entrypoint:

https://docs.github.com/en/rest/actions/self-hosted-runners#create-a-registration-token-for-a-repository

**Reply by AF250329** [reply]
Mmmm.... The problem is that this is Organization repository and not my own :cry: I am just an `admin` there and I have not rights (besides some basic/viewer or something) in Organization. So when I use `/repos` entrypoint I get `Not found` error

**I guess it is a bug 🐛 in a Github**. Some users with some access rights in repository could actually **see** the token in 'instructions' page but **could not** generate it via the API calls. In my case I am `admin` in this repository but not an `admin:org` - so I could not generate token via API call.... I think this is a bug !

**Reply by AF250329** [reply]
I can generate token for **my own** repositories that is my **private repo** with `/repos` entrypoint

**Reply by jsoref** [reply]
Afaiu, you should be able to create a token with the repos API. Open a ticket to support.

**Reply by AF250329** [reply]
Here I am trying to generate token for repository in Organization:

![image](https://user-images.githubusercontent.com/78792938/196036484-c5a4d0d1-774a-46b8-a79f-6c36d9ca5b7d.png)

Here I am trying to generate token for repository (only name of repository)

![image](https://user-images.githubusercontent.com/78792938/196036520-ca531b91-f436-415d-9d51-e990d5830655.png)


I am trying to generate token for repository by using organization name and repository name

![image](https://user-images.githubusercontent.com/78792938/196036551-61e812d9-b6e3-4351-84d7-e5065946752d.png)


Any help ?

**Reply by jsoref** [reply]
You're doing _something_ wrong.

This worked:
```
gh api /repos/jsoref/examples-testing/actions/runners/registration-token --method POST\
```

This failed w/a 403:
```
gh api /repos/powerdns/pdns/actions/runners/registration-token --method POST
{
  "message": "Must have admin rights to Repository.",
  "documentation_url": "https://docs.github.com/rest/reference/actions#create-a-registration-token-for-a-repository"
}
```

You need to remove `AF250329` -- you aren't trying to create a runner in your space, you're trying to create it in a different space.

**Reply by AF250329** [reply]
`examples-testing` --> it is **your** private repo - it does not created under organization `powerdns`. I could generate tokens for **my private** repo without any problems

**Reply by jsoref** [reply]
Right, but the point is that your third example was broken, it had your username before your organization, it should only have your organization. I don't think I'm an admin to any repos for which I'm not also an org owner....

**Reply by jsoref** [reply]
Btw, it's a *lot* easier if you copy and paste text instead of taking screen captures -- it's easier to censor them too. Your font contrast is painful for some people's eyes and if I were using a screen reader, I wouldn't be able to help.

To censor things, you can do something like:
```PS
$Env:org='powerdns'
$Env:repo='pdns'
gh api /repos/$Env:org/$Env:repo/actions/runners/registration-token --method POST
{
  "message": "Must have admin rights to Repository.",
  "documentation_url": "https://docs.github.com/rest/reference/actions#create-a-registration-token-for-a-repository"
}
gh: Must have admin rights to Repository. (HTTP 403)
```

And just not include the first lines...

**Comment by AF250329** [comment, marked as answer]
Thanks to @jsoref I found a solution.  May be this is a bug in Github documentation, may be this is a bug in my Organization setup - who knows but this is what you need to do:

According to Github documentation:  [Create a registration token for a repository](https://docs.github.com/en/rest/actions/self-hosted-runners#create-a-registration-token-for-a-repository)  when you trying to generate token for **repository** you must send those parameters:

```powershell
gh api \
  --method POST \
  -H "Accept: application/vnd.github+json" \
  /repos/OWNER/REPO/actions/runners/registration-token
```
Path parameters

- **OWNER**            The account owner of the repository. The name is not case sensitive.
- **REPO**                The name of the repository. The name is not case sensitive.


but in my case I changed `owner` to my organization name (instead of my username) and voila ! it worked !!!

Thanks again @jsoref

**Reply by ti3z9w1-BRONGA** [reply]
Do I understand that you have the Github CLI (gh) running on a self-hosted runner?  How did you manage that?  None of our self-hosted runners have it installed, and I've only found 1 mechanism in the Marketplace for installing it, but we host github in-house, and they have not as yet allowed access to that particular Marketplace item.

Thanks!
